client/modules/cm-tarnation/src/host.ts

```diff
@@ -0,0 +1,324 @@
+        const buffer = this.language.cache.find(f.tree, start, f.to)
+        if (buffer) {
+          // try to find a suitable chunk from the buffer to restart the tokenizer from
+          const { chunk, index } = buffer.search(this.region.edit!.from, -1)
+          if (chunk && index !== null) {
+            // split the buffer, reuse the left side,
+            // but keep the right side around for reuse as well
+            const { left, right } = buffer.split(index)
+            this.tokenizerPreviousRight = right
+            this.region.from = chunk.context.pos
+            this.setupTokenizer(left, chunk.context)
+
+            // check if parser has a cached state for this chunk
+            if (this.language.cache.has(chunk)) {
+              const context = ParserContext.deserialize(this.language.cache.get(chunk)!)
+              this.setupParser(context)
+            }
+          }
+        }
+      }
+    }
+
+    // if we couldn't reuse state, we'll need to startup things with a default state
+    if (!this.tokenizer) this.setupTokenizer()
+    if (!this.parser) this.setupParser()
+  }
+
+  /**
+   * Instantiates the `Tokenizer`.
+   *
+   * @param buffer - A `TokenizerBuffer` to reuse.
+   * @param context - A `TokenizerContext` to reuse.
+   */
+  private setupTokenizer(buffer?: TokenizerBuffer, context?: TokenizerContext) {
+    if (!buffer || !context) {
+      const stack = new TokenizerStack({ stack: [["root", {}]], embedded: null })
+      context = new TokenizerContext(this.region.from, stack)
+      buffer = new TokenizerBuffer()
+    }
+
+    this.tokenizer = new Tokenizer(
+      this.language,
+      context,
+      buffer,
+      this.input,
+      this.region
+    )
+  }
+
+  /**
+   * Instantiates the `Parser`.
+   *
+   * @param context - A `ParserContext` to reuse.
+   */
+  private setupParser(context?: ParserContext) {
+    if (!context) {
+      context = new ParserContext(this.start, 0, new ParserBuffer(), new ParserStack(), {
+        pending: [],
+        parsers: []
+      })
+    }
+
+    this.parser = new Parser(
+      this.language,
+      context,
+      this.input,
+      this.region,
+      [],
+      this.context
+    )
+  }
+
+  /**
+   * The current "position" of the host. This isn't really all that
+   * accurate, as it's only reporting the tokenizer's position. That means
+   * when the parser is running, the position will just be sitting still.
+   */
+  get pos() {
+    return this.tokenizer.context.pos
+  }
+
+  /** Advances the tokenizer or parser one step, depending on the current stage. */
+  advance(): Tree | null {
+    if (!this.measurePerformance) this.measurePerformance = perfy()
+    switch (this.stage) {
+      case Stage.Tokenize: {
+        // try to reuse ahead state
+        const reused =
+          this.tokenizerPreviousRight &&
+          this.tokenizer.tryToReuse(this.tokenizerPreviousRight)
+
+        // can't reuse the buffer more than once (pointless)
```

Comment:
Why is this? Out of curiosity?

Reply:
The reasoning is really simple, actually. The tokenizer emits to a buffer which holds a bunch of chunks. If we reuse another buffer on the right (as in, we concatenate it), we just move all the way to the end of the newly concatenated buffer. We reuse the entire thing, so there isn't any point in checking if we can still reuse it. We already did!